File: hw1/HW1Tests/ParserTests.cs
using hw1;
using Xunit;

namespace HW1Tests
{
    public class ParserTests
    {
        private const int Correct = 0;
        private const int WrongArgCount = 1;
        private const int WrongArgFormat = 2;
        private const int WrongOperation = 3;
        
        [Theory]
        [InlineData("+", Operations.Plus)]
        [InlineData("-", Operations.Minus)]
        [InlineData("*", Operations.Multiply)]
        [InlineData("/", Operations.Divide)]
        public void TryParseArguments_Return_Correct(string operation, Operations operationExpected)
        {
            var args = new[] { "6", operation, "35" };
            var isValid = Parser.TryParseArguments(args, out var operationActual, out var val1, out var val2);
            Assert.Equal(Correct, isValid);
            Assert.Equal(6, val1);
            Assert.Equal(operationExpected, operationActual);
            Assert.Equal(35, val2);
        }
        
        [Fact]
        public void TryParseArguments_WrongCount_WillReturnWrongArgCount()
        {
            var args = new[] { "7", "+"};
            var check = Parser.TryParseArguments(args, out _, out _, out _);
            Assert.Equal(WrongArgCount, check);
        }
        
        [Fact]
        public void TryParseArguments_NotInt_WillReturnWrongArgFormat()
        {
            var args = new[] { "7", "*", "4.6" };
            var isValid = Parser.TryParseArguments(args, out _, out _, out _);
            Assert.Equal(WrongArgFormat, isValid);
        }
        
        [Fact]
        public void TryParseArguments_WrongOperation_WillReturnWrongOperation()
        {
            var args = new[] { "51", "%", "25" };
            var check = Parser.TryParseArguments(args, out _, out _, out _);
            Assert.Equal(WrongOperation, check);
        }
    }
}

File: HW4/HW4Test/ParserTests.fs
module Tests.ParserTests

open System
open HW4
open Xunit

[<Theory>]
[<InlineData("+", Operations.Plus)>]
[<InlineData("-", Operations.Minus)>]
[<InlineData("*", Operations.Multiply)>]
[<InlineData("/", Operations.Divide)>]
let ``tryParseArgs_Operation_Parse_Correct`` (operation, operationExpected) =
    let args = [|"435";operation;"113"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Operations.Plus
    let check = Parser.tryParseArgs args &operationResult &val1 &val2
    Assert.Equal(ErrorCodes.Correct, check)
    Assert.Equal(435, val1)
    Assert.Equal(operationExpected, operationResult)
    Assert.Equal(113, val2)
    
[<Fact>]
let ``tryParseArgs_NotNumber_WillReturnWrongArgFormatInt`` () =
    let args = [|"4";"+";"3.6"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Operations.Plus
    let check = Parser.tryParseArgs args &operationResult &val1 &val2
    Assert.Equal(ErrorCodes.WrongArgFormat, check)
    
[<Fact>]
let ``tryParseArgs_NotOperation_WillReturnWrongOperation`` () =
    let args = [|"21";"?";"37"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Operations.Plus
    let check = Parser.tryParseArgs args &operationResult &val1 &val2
    Assert.Equal(ErrorCodes.WrongOperation, check)

[<Fact>]
let ``tryParseArgs_WrongArgsCount_WillReturnWrongArgCount`` () =
    let args = [|"11";"*";"142";"dds"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Operations.Plus
    let check = Parser.tryParseArgs args &operationResult &val1 &val2
    Assert.Equal(ErrorCodes.WrongArgCount, check)    